src/main/scala/cognite/spark/v1/AssetHierarchyBuilder.scala

```diff
@@ -69,6 +69,26 @@ final case class InvalidNodeReferenceException(nodeIds: Seq[String], referencedF
       val refNodes = referencedFrom.map(x => s"'$x'").sorted.mkString(", ")
       s"${plural("Parent", "Parents")} $nodes referenced from $refNodes ${plural("does", "do")} not exist."
     })
+final case class AttemptedRootChangeException(cause: CdpApiException, newSubtreeRoot: Asset)
+    extends CdfSparkException(
+      {
+        val commaSeparatedAssets =
+          cause.duplicated.get
+            .map {
+              _.toIterable
+                .map { case (key, value) => s"$key=${value.asString.getOrElse(value.toString)}" }
+                .mkString("(", ", ", ")")
+            }
+            .mkString(", ")
+
+        val withRootAssetId = newSubtreeRoot.rootId.map(id => s" with id $id").getOrElse("")
```

Comment:
Maybe display externalId as well, if set.

Reply:
Unfortunately, we don't know the externalId of the root asset without making an extra call to CDF, as the `newSubtreeRoot` is not necessarily an actual root asset, just the "uppermost" node specified in the input data. CDF gives us a `rootId`, but not a `rootExternalId`, afaik.

Reply:
Yeah, could make such a call where we construct this exception, but perhaps not worth the effort.